hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/index/bloom/HoodieMetadataBloomIndexCheckFunction.java

```diff
@@ -0,0 +1,166 @@
+    return new BloomIndexLazyKeyCheckIterator(tuple2Iterator);
+  }
+
+  private class BloomIndexLazyKeyCheckIterator extends LazyIterableIterator<Tuple2<String, HoodieKey>, List<HoodieKeyLookupResult>> {
+    public BloomIndexLazyKeyCheckIterator(Iterator<Tuple2<String, HoodieKey>> tuple2Iterator) {
+      super(tuple2Iterator);
+    }
+
+    @Override
+    protected void start() {
+    }
+
+    @Override
+    protected List<HoodieKeyLookupResult> computeNext() {
+      // Partition path and file name pair to list of keys
+      final Map<Pair<String, String>, List<HoodieKey>> fileToKeysMap = new HashMap<>();
+      final Map<String, HoodieBaseFile> fileIDBaseFileMap = new HashMap<>();
+      final List<HoodieKeyLookupResult> resultList = new ArrayList<>();
+
+      while (inputItr.hasNext()) {
+        Tuple2<String, HoodieKey> entry = inputItr.next();
+        final String partitionPath = entry._2.getPartitionPath();
+        final String fileId = entry._1;
+        if (!fileIDBaseFileMap.containsKey(fileId)) {
+          Option<HoodieBaseFile> baseFile = hoodieTable.getBaseFileOnlyView().getLatestBaseFile(partitionPath, fileId);
+          if (!baseFile.isPresent()) {
+            throw new HoodieIndexException("Failed to find the base file for partition: " + partitionPath
+                + ", fileId: " + fileId);
+          }
+          fileIDBaseFileMap.put(fileId, baseFile.get());
+        }
+        fileToKeysMap.computeIfAbsent(Pair.of(partitionPath, fileIDBaseFileMap.get(fileId).getFileName()),
+            k -> new ArrayList<>()).add(entry._2);
+        if (fileToKeysMap.size() > BLOOM_FILTER_CHECK_MAX_FILE_COUNT_PER_BATCH) {
+          break;
+        }
+      }
+      if (fileToKeysMap.isEmpty()) {
+        return Collections.emptyList();
+      }
+
+      List<Pair<String, String>> partitionNameFileNameList = fileToKeysMap.keySet()
+          .stream().collect(Collectors.toList());
+      Map<Pair<String, String>, ByteBuffer> fileIDToBloomFilterByteBufferMap =
+          hoodieTable.getMetadataTable().getBloomFilters(partitionNameFileNameList);
+
+      final AtomicInteger totalKeys = new AtomicInteger(0);
+      fileToKeysMap.forEach((partitionPathFileNamePair, hoodieKeyList) -> {
+        final String partitionPath = partitionPathFileNamePair.getLeft();
+        final String fileName = partitionPathFileNamePair.getRight();
+        final String fileId = FSUtils.getFileId(fileName);
+        ValidationUtils.checkState(!fileId.isEmpty());
+
+        if (!fileIDToBloomFilterByteBufferMap.containsKey(partitionPathFileNamePair)) {
+          throw new HoodieIndexException("Failed to get the bloom filter for " + partitionPathFileNamePair);
+        }
+        final ByteBuffer fileBloomFilterByteBuffer = fileIDToBloomFilterByteBufferMap.get(partitionPathFileNamePair);
+
+        HoodieDynamicBoundedBloomFilter fileBloomFilter =
```

Comment:
metadata table always used Dynamic bloom is it?

Reply:
HUDI-3203 will address this. It will be based on the type we read from the index payload